### PR TITLE
Fix cyclic CMake dependency on Windows with BUILD_SEPARATE_OPS=ON

### PR DESCRIPTION
**Pull request by xuhancn**
### ❌ BEFORE: Cyclic Dependency

```
torch_xpu.dll ── target_link_libraries( PUBLIC torch_xpu_ops) ──▶ torch_xpu_ops (STATIC)
      ▲                                                                    │
      │          setup_common_libraries():                                 │
      │            target_link_libraries(torch_xpu_ops PUBLIC torch_xpu)   │
      └────────────────────────────────────────────────────────────────────┘
                                     CYCLE!

Kernel DLL cross-deps: file(GET_RUNTIME_DEPENDENCIES .../kernel.dll ...) — hardcoded file paths
Linker: /FORCE:UNRESOLVED — ignores missing symbols (dangerous hack)
```

### ✅ AFTER: xpu_hal.dll (Function Pointer Bridge in torch-xpu-ops)

```
c10.dll (always linked, pure SYCL)

torch_xpu.dll ──▶ xpu_hal.dll (SHARED, only links c10, no reverse edge!)
      │                 ▲
      │                 │
      │          kernel DLLs (SHARED, ~180, WINDOWS_EXPORT_ALL_SYMBOLS)
      │                 │
      └─────────────────┘
      target_link_libraries(PRIVATE ...) via cmake targets

torch_xpu_ops (STATIC) ──▶ torch_cpu, c10  (no torch_xpu dep!)
      │
      └── NO /FORCE:UNRESOLVED
```

### Changes

| File | What |
|------|------|
| **xpu_hal.dll** | `src/hal/XPUHal.{h,cpp}` — function pointer bridge for generator ops |
| **Delete cycle edge** | `src/BuildOnWindows.cmake` — removed `target_link_libraries(torch_xpu_ops PUBLIC torch_xpu)` |
| **xpu_hal → Caffe2Targets** | `src/BuildOnWindows.cmake` — `install(TARGETS xpu_hal EXPORT Caffe2Targets ...)` fixes CI cmake export error |
| **Kernel DLLs link xpu_hal** | `src/BuildOnWindows.cmake` — `target_link_libraries(kernel_dll PRIVATE xpu_hal)` |
| **Remove /FORCE:UNRESOLVED** | `src/BuildOnWindows.cmake` — no longer needed |
| **WINDOWS_EXPORT_ALL_SYMBOLS** | `src/BuildOnWindows.cmake` — kernel DLLs export SYCL entry points |
| **Generator → xpu_hal** | 16 kernel files — `xpu_hal::getDefaultGenerator()` / `xpu_hal::philoxState()` |
| **DeviceProperties → raw SYCL** | `src/comm/DeviceProperties.h` — uses `c10::xpu::get_raw_device()` |
| **__has_include backcompat** | 4 kernel files — backwards compatibility with older pytorch without xpu_hal |
| **Cross-kernel deps** | `src/BuildOnWindows.cmake` — BatchNorm→Resize, UnaryComplex→Copy/Sign, Pow→Unary/UnaryFraction, Shape→ShapeKernels |
| **Same-module deps** | `src/BuildOnWindows.cmake` — ReduceOps→specializations, Indexing, Distribution, BinaryDiv, Chebyshev, Embedding, Dropout/FusedAdam/GridSampler, Histogram, Foreach ternary |

### Removed
- `/FORCE:UNRESOLVED` linker flag — all symbols properly resolved
- Hardcoded `file(GET_RUNTIME_DEPENDENCIES ...)` file paths
- `target_link_libraries(torch_xpu_ops PUBLIC torch_xpu)` — the cycle edge

### Build Verified
```
BUILD_SEPARATE_OPS=ON → 0 compile errors ✅
BUILD_SEPARATE_OPS=OFF → same pre-existing AOT limitation (ocloc not available)
Runtime: torch.rand, torch.randn, torch.randint with Generator("xpu") → all pass ✅
```

### CODE REVIEW

**Comment by xuhancn** [comment]
Closing this PR as the architecture has changed significantly.

**Situation:**
- This PR was based on an older torch-xpu-ops architecture that included `src/hal/XPUHal.cpp` and `src/hal/XPUHal.h`
- The current main branch (commit `dbb06c23...`) has removed the hal/ directory entirely as part of an architectural refactor
- This PR adds files and modifications that no longer apply to the current codebase

**Issue #3209 Status:**
As noted in the issue comments, BUILD_SEPARATE_OPS on Windows is a corner case that is not guaranteed to work. If the cyclic dependency issue persists in the new architecture and needs fixing, a new solution should be designed that's compatible with the current codebase structure.

**Related PR:**
PyTorch PR #188284 has been updated to fetch from intel/torch-xpu-ops (commit `dbb06c23...`) and is currently in CI.

**Comment by xuhancn** [comment]
## PR Reopened - Architecture Compatibility Issue

⚠️ **Important**: This PR has compatibility issues with the current codebase.

**Problem**:
- This PR adds `src/hal/XPUHal.cpp` and `src/hal/XPUHal.h`
- The current main branch (commit `dbb06c23...`) **does not have** a `src/hal/` directory
- The hal/ architecture was removed in a recent refactor

**Current PR Status**:
- Based on: Old architecture with hal/ bridge pattern
- Target: Current main without hal/ directory
- Result: **Cannot merge without conflicts**

**Options**:
1. **Close this PR** - Recommended if BUILD_SEPARATE_OPS cyclic deps are resolved in current architecture
2. **Rewrite the fix** - Adapt the solution to work with current codebase structure (no hal/ directory)
3. **Keep open for reference** - Document the old approach for future investigation

**Related**:
- Issue #3209 remains open (BUILD_SEPARATE_OPS on Windows is a corner case)
- PyTorch PR #188284 updated to use intel/torch-xpu-ops main (without hal/ code)

**Comment by xuhancn** [comment]
Closing this PR in favor of #4200.

**Why the change:**
- PR #4129 was based on an old fork branch (`xuhancn:fix/issue-3209-fix-cyclic-cmake-dependency`)
- PR #4200 is a fresh branch based on current `intel:main` with the same patches applied cleanly

Both PRs contain identical code, but #4200 has a cleaner git history and is directly in the intel repo.
